index: Fast path for EmptyPostings cases in Merge, Intersect and Without

bwplotka opened the pull request:
Updates old https://github.com/prometheus/tsdb/pull/233 onto fresh TSDB code.

closes https://github.com/prometheus/tsdb/pull/233

bwplotka commented:
ping @gouthamve

krasi-georgiev commented:
@bwplotka 	I did a PR against your branch with few adjustments.
https://github.com/bwplotka/tsdb/pull/2

krasi-georgiev commented:
LGTM, Thanks!

krasi-georgiev commented:
@codesome 	want to have a quick look before I merge it?

krasi-georgiev commented:
ping @bwplotka

bwplotka commented:
Updated @krasi-georgiev and @codesome 

Thanks for ping & reviews. Ping me directly on IRC or slack if you want my review/PR update I am flooded with GH notifications ):

I changed a lot now, because of conflicts - I rewrote test cases to make more sense, let me know if that is OK with you guys.

krasi-georgiev commented on index/postings.go:
```diff
@@ -395,18 +409,24 @@ type mergedPostings struct {
-	return &mergedPostings{h: ph}
+
+	if len(ph) == 0 {
+		return nil, false
```
shouldn't this be before anything else?

bwplotka replied:
hm? We know this only after `it.Next` bit.

krasi-georgiev replied:
yep , I missed that bit.

krasi-georgiev commented on index/postings_test.go:
```diff
@@ -61,39 +61,125 @@ func TestMemPostings_ensureOrder(t *testing.T) {
 	for _, c := range cases {
-		a := newListPostings(c.a)
-		b := newListPostings(c.b)
+		t.Run("", func(t *testing.T) {
```
maybe add some indication to know when a test fails which case failed.
```
map[string]struct {
.
.

for t, c := range cases {
		t.Run(t, func(t *testing.T) {
```

bwplotka replied:
not needed. Passing `""` cause go test to put a number, so exactly what you propose.

krasi-georgiev replied:
yes that is right but when you have many cases it is kind of difficult to count which one failed exactly.

bwplotka replied:
Disagree, let's not obfuscate the tests more -> everyone can count

krasi-georgiev replied:
that is why I added the title in:
https://github.com/prometheus/tsdb/pull/559/files

bwplotka replied:
Yea, I agree there is a value for this, but I am afraid I won't have time to do this, this week ): 

I would suggesting merging this PR with just indexing (numbers), then we can work on making it more verbose later on. What do you think?

krasi-georgiev replied:
sure I don't mind can always add it later if it gets annoying to count :smile:

krasi-georgiev commented on index/postings_test.go:
```diff
@@ -61,39 +61,125 @@ func TestMemPostings_ensureOrder(t *testing.T) {
+
+			if i == EmptyPostings() {
+				t.Fatal("intersect unexpected result: EmptyPostings sentinel")
+			}
```
why do we need?
```
if c.res == EmptyPostings() {
	testutil.Equals(t, EmptyPostings(), i)
	return
}

if i == EmptyPostings() {
	t.Fatal("intersect unexpected result: EmptyPostings sentinel")
}
```

the test passes even without these extra checks.

bwplotka replied:
Exactly to expect `EmptyPostings` explicitly!

krasi-georgiev replied:
yep, now I see it.

krasi-georgiev commented on index/postings_test.go:
```diff
@@ -541,3 +698,81 @@ func TestIntersectWithMerge(t *testing.T) {
+
+	for _, c := range cases {
+		t.Run("", func(t *testing.T) {
+			if c.res == nil {
```
don't think we need this. A panic is ok in a test in my opinion

bwplotka replied:
If you think so, can remove

krasi-georgiev commented on index/postings_test.go:
```diff
@@ -197,35 +283,106 @@ func TestMultiMerge(t *testing.T) {
+		},
+		{
+			in: []Postings{
+				toPostings(1, 2, 3, 4, 5),
```
what is blocking to use `newListPostings` everywhere?

bwplotka replied:
boilerplate

krasi-georgiev replied:
how about if you change 
`newListPostings([]uint64) -> newListPostings(list ...uint64)`
than you can call it as `newListPostings(1, 2, 3, 4, 5)`
just tried it and it removes that boilerplate without the need for an extra func.


not a blocker though can leave it as if you prefer.

bwplotka replied:
It will increase this PR size, but did it, I like this idea :+1:

krasi-georgiev commented on index/postings.go:
```diff
@@ -382,13 +388,21 @@ func Merge(its ...Postings) Postings {
-// are not in the drop list
+// are not in the drop list.
 func Without(full, drop Postings) Postings {
+	if full == EmptyPostings() {
```
nope, drop are the ones to be removed

krasi-georgiev commented:
Thanks Bartek!

bwplotka commented:
Thanks Krasi (:
